src/solvers/refinement/string_constraint_generator_valueof.cpp

```diff
@@ -207,64 +202,41 @@ string_exprt string_constraint_generatort::add_axioms_from_int_with_radix(
     not_exprt(equal_exprt(res[1], zero_char)));
   axioms.push_back(a4);
 
-  for(size_t size=1; size<=max_size; size++)
+  for(size_t j=1; j<max_size; ++j)
   {
-    // For each possible size, we have:
-    // sum_0 = numeric value of res[0] (which is 0 if res[0] is '-')
-    // sum_j = radix * sum_{j-1} + numeric value of res[j]
-    // and add axioms:
-    // a5 : |res| == size =>
-    //        forall 1 <= j < size.
-    //          is_number && (j >= max_size-2 => no_overflow)
-    //     where is_number := res[j] is a digit for the radix
-    //     and no_overflow := sum_{j-1} = (radix * sum_{j-1} / radix)
-    //                        && sum_j >= sum_{j - 1}
-    //         (the first part avoid overflows in multiplication and
-    //          the second one in additions)
-    // a6 : |res| == size && res[0] is a digit for radix => input_int == sum
-    // a7 : |res| == size && res[0] == '-' => input_int == -sum
-
-    exprt::operandst digit_constraints;
-    exprt sum=
-      get_numeric_value_from_character(res[0], char_type, type, radix_ul);
-
-    for(size_t j=1; j<size; j++)
-    {
-      mult_exprt radix_sum(sum, radix_input_type);
-      // new_sum = radix * sum + (numeric value of res[j])
-      exprt new_sum=plus_exprt(
-        radix_sum,
-        get_numeric_value_from_character(res[j], char_type, type, radix_ul));
-      exprt is_number=
-        is_digit_with_radix_lower_case(res[j], radix_char_type, radix_ul);
-      digit_constraints.push_back(is_number);
-
-      // An overflow can happen when reaching the last index of a string of
-      // maximal size which is `max_size-1` for negative numbers and
-      // `max_size - 2` for positive numbers because of the absence of a `-`
-      // sign.
-      if(j>=max_size-2)
-      {
-        and_exprt no_overflow(
-          equal_exprt(sum, div_exprt(radix_sum, radix_input_type)),
-          binary_relation_exprt(new_sum, ID_ge, radix_sum));
-        digit_constraints.push_back(no_overflow);
-      }
-      sum=new_sum;
-    }
-
-    equal_exprt premise=res.axiom_for_has_length(size);
-    implies_exprt a5(premise, conjunction(digit_constraints));
-    axioms.push_back(a5);
-
-    implies_exprt a6(
-      and_exprt(premise, starts_with_digit), equal_exprt(input_int, sum));
-    axioms.push_back(a6);
+    /// We need to constrain res[j] to be a valid digit for this radix and for
+    /// all j<res.length()
+    const if_exprt chr(
+      binary_relation_exprt(from_integer(j, index_type), ID_lt, res.length()),
+      res[j],
+      zero_char);
+    axioms.push_back(
+      is_digit_with_radix_lower_case(chr, radix_char_type, radix_ul));
```

Comment:
can we pass the call inside the `if`? What I mean is `j<res.length?is_digit(chr):true` instead of `is_digit(j<res.length?chr:'0')`. It seems to me the generated formula would be slightly simpler.